### DISCUSSION TITLE
Concurrent invokes?

### DISCUSSION THREAD

**Discussion by aktxyz** [jamesmh/coravel#262] [Q&A]
If I am have an every-second schedule, and at some time the Invoke() method takes longer than 1 second, does that mean Invoke() may be called again before the previous one finishes?

I guess another way of asking, is the every-second interval wall time? or 1 second since the previous invoke returned?

**Comment by jamesmh** [comment]
Yes that is what will happen - both tasks will be running at the same time.

You can prevent this by using the `PreventOverlap()` feature:

[https://docs.coravel.net/Scheduler/#prevent-overlapping-tasks](https://docs.coravel.net/Scheduler/#prevent-overlapping-tasks)

**Reply by mkojoa** [reply]
Does this also happens for queued listeners?

**Reply by mkojoa** [reply]
At the moment I have event with a queued listener that handle generating payslip-report and sending emails when user clickes on "send payslip by mail" , everything runs fine but every user has to wait until the current running task is completed before the server starts a new task. Since they can take several minutes to complete this can represent a problem.

How can I execute multiple background tasks at the same time?

**Reply by jamesmh** [reply]
So event listeners run on the thread that you broadcasted it on. If you queue it, it runs on one queue (concurrent queues are not available yet). [Internally it uses `WhenAll()`](https://github.com/jamesmh/coravel/blob/6f11cda3ea7cc7b5a16e7cb4edc39176488cc9d2/Src/Coravel/Queuing/Queue.cs#L102) so async operations may run concurrently, but CPU-heavy operations will not.

**Reply by mkojoa** [reply]
So if the listener is not queued and requires cup-heavy operations what happens? No concurrency?


When a "send payslip by email" button is clicked multiple times, and the first click process gets started what happens to the second click processing?

**Reply by jamesmh** [reply]
1. Listener is not queued then the work happens on the thread that initiated the event broadcast. For example, if this was done from a web controller request thread and you broadcasted an event, and the work inside the listener took 30 seconds to process, then your web request would be "hanging" there for 30 seconds until the listener is complete - then your response will be sent to the HTTP client. This is why queuing work is generally preferred when doing work in web contexts.

2. 

> When a "send payslip by email" button is clicked multiple times, and the first click process gets started what happens to the second click processing?

That depends on how you've coded your application. Are you blocking concurrent clicks? Are you allowing them? Are you rate-limiting at all? Are you using dedupe mechanism to detect operations that you be idempotent / blocked? 

From Coravel's point of view, if your code queues work then it will _do_ the work your code asked it to do. You are responsible for determining whether work should be queued or not.

**Reply by mkojoa** [reply]
Thanks, @jamesmh, the explanation was clear. 

The diagram below shows what I am trying to achieve with coravel event queuing.

I just need to find my way out.

![multiple-queue](https://user-images.githubusercontent.com/20341997/210359257-e7fdab5b-5fba-43fa-be25-cc85ac73ae86.png)

**Comment by aktxyz** [comment]
wow ! that's perfect
